File: frontend/src/lib/api/repoDashboard.ts
import { publicApiBaseUrl } from '$lib/api/config';

async function readJson<T>(res: Response): Promise<T> {
	if (!res.ok) {
		const text = await res.text();
		throw new Error(text || `HTTP ${res.status}`);
	}
	return (await res.json()) as T;
}

export type DashboardSummary = {
	specs_tracked: number;
	latest_commit_sha: string | null;
	latest_commit_message: string | null;
	latest_commit_at: string | null;
	coverage_percent: number | null;
	missing_items: number;
	low_confidence_items: number;
	candidate_artifacts: number;
};

export type DashboardSpecRow = {
	spec_id: number;
	paper_id: string;
	latest_version_id: number;
	latest_commit_sha: string;
	status: string;
	satisfied: number;
	missing: number;
	low_confidence: number;
	candidate_artifacts: number;
	last_evaluated_commit_sha: string | null;
	last_evaluated_at: string | null;
};

export type DashboardAttentionItem = {
	spec_id: number;
	paper_id: string;
	headline: string;
	detail: string;
	severity: string;
};

export type DashboardRecentChange = {
	spec_id: number;
	paper_id: string;
	local_key: string | null;
	message: string;
	commit_sha: string;
	committed_at: string;
};

export type DashboardArtifactActivity = {
	filepath: string;
	link_type: string;
	item_count: number;
	spec_paper_id: string | null;
	label: string;
};

export type RepoDashboardResponse = {
	repo_id: number;
	repo_name: string;
	summary: DashboardSummary;
	specs: DashboardSpecRow[];
	needs_attention: DashboardAttentionItem[];
	recent_changes: DashboardRecentChange[];
	artifact_activity: DashboardArtifactActivity[];
};

export type SidebarSpecVersion = {
	version_id: number;
	commit_sha: string;
	commit_message: string;
	committed_at: string;
	title: string | null;
	status: string;
};

export type SidebarSpec = {
	id: number;
	paper_id: string;
	title: string;
	versions: SidebarSpecVersion[];
};

export type RepoSidebarResponse = {
	specs: SidebarSpec[];
};

export async function fetchRepoDashboard(repoId: string | number): Promise<RepoDashboardResponse> {
	const res = await fetch(`${publicApiBaseUrl()}/repos/${repoId}/dashboard`);
	return readJson(res);
}

export async function fetchRepoSidebar(repoId: string | number): Promise<RepoSidebarResponse> {
	const res = await fetch(`${publicApiBaseUrl()}/repos/${repoId}/sidebar`);
	return readJson(res);
}

File: api/src/specvsreality_api/schemas/repo_dashboard.py
"""Response schemas for the repository dashboard."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class DashboardSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    specs_tracked: int
    latest_commit_sha: str | None
    latest_commit_message: str | None
    latest_commit_at: datetime | None
    coverage_percent: float | None
    missing_items: int
    low_confidence_items: int
    candidate_artifacts: int


class DashboardSpecRow(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec_id: int
    paper_id: str
    latest_version_id: int
    latest_commit_sha: str
    status: str
    satisfied: int
    missing: int
    low_confidence: int
    candidate_artifacts: int
    last_evaluated_commit_sha: str | None
    last_evaluated_at: datetime | None


class DashboardAttentionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec_id: int
    paper_id: str
    headline: str
    detail: str
    severity: str


class DashboardRecentChange(BaseModel):
    model_config = ConfigDict(extra="forbid")

    spec_id: int
    paper_id: str
    local_key: str | None
    message: str
    commit_sha: str
    committed_at: datetime


class DashboardArtifactActivity(BaseModel):
    model_config = ConfigDict(extra="forbid")

    filepath: str
    link_type: str
    item_count: int
    spec_paper_id: str | None
    label: str


class RepoDashboardResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    repo_id: int
    repo_name: str
    summary: DashboardSummary
    specs: list[DashboardSpecRow]
    needs_attention: list[DashboardAttentionItem]
    recent_changes: list[DashboardRecentChange]
    artifact_activity: list[DashboardArtifactActivity]
